### DISCUSSION TITLE
集中化管理mpv的脚本选项

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#126] [INFO 信息]
[**script-opts.conf**](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/script-opts.conf)
<s>_可能以后会用这种方式管理脚本选项，也可能不会。_
目前就当作是水一帖，分享一种花式使用mpv的方法吧。</s>

核心参数就是被我嫌弃的 [`--script-opts`](https://mpv.io/manual/master/#options-script-opts) _这个选项的用法介绍见下方（该展示的部分在当前的主设置文件中已被移除）_
https://github.com/hooke007/mpv_PlayKit/blob/fb819f16a5ae2cccf3913df10a1ccfa3a77fa72d/portable_config/mpv.conf#L206-L211

如果你用的脚本不多、你要改的脚本选项不多、部分libmpv前端的脚本兼容做的不完善，这些情况下显然 `--script-opts` 方便快捷。
相反假设你的脚本选项很多，用单项 `--script-opts` 罗列全部显然不现实，用多项 `--script-opts-append` 一一列举也会把你的 mpv.conf 塞得很长显得乱，这时候对每个脚本单独设置 conf 文件在条理上更清晰。

**这两种方法是可以同时使用的，当然别忘了遇到重名的选项， --script-opts 会覆盖你 脚本conf 文件里的设置**

在懒人包里我分离了各种 配置预设 到单独的 profiles.conf 文件，主要目的是规避 <s>我偶尔眼瞎漏看header</s> 萌新不熟悉语法踩坑，但同时也是为了保持 mpv.conf 主体部分的简洁。基于同样的思路，启发了我用附属设置文件管理脚本选项的想法。

[`--include`](https://mpv.io/manual/master/#options-include) 这个选项也是比较特殊，可以使用多次来指定不同的附属设置文件。
```ini
# 随便写的mpv.conf
profile=11
include=~~/aaa.conf
profile=22
include=~~/bbb.conf
profile=33
include=~~/ccc.conf
```
在主设置的全部 **其它** 参数读取完之后，再按顺序依次读取 aaa bbb ccc 中的选项。

一些提示：
- <s>因为暂不打算用这种方式取代，所以</s>没有做全面测试，涉及字符串相关的脚本选项，可能存在语法上的暗病。
- 它此时作为 mpv.conf 的一部分，因此格式相对的自由没有脚本设置文件中的那些限制，排版也对阅读来说更友好（就是选项名有一点点长）

**Comment by 921j** [comment]
请教下大佬，如果我想把 mpv.conf、input_uosc.conf 里的自定义项保存在 script-opts.conf 里该怎么办呢？
比如 border = no 这项，我尝试了以下几条配置貌似都没有作用

 script-opts-append = osc-windowcontrols=no
 script-opts-append = border=no
 script-opts-append = mpv-border=no

我希望以后更新 mpv-lazy 的时候只需要把旧配置里的脚本及配置复制过来，然后对比修改新的 script-opts.conf 就行了，不需要在一一修改 mpv.conf、input_uosc.conf 以及 script-opts 里各个脚本的配置，请问能做到吗？

**Reply by hooke007** [reply]
首先区分 mpv.conf input.conf
- input_uosc.conf 依然是一种 input.conf 不可能保存在 mpv.conf 里
- script-opts.conf profiles.conf 也是 mpv.conf ，前两者的参数可以合并进 mpv.conf
- `script-opts-append` 是属于 mpv 的选项；`border` 也是属于 mpv 的选项。我不理解你串联两个 mpv选项 的目的。直接分开写不行吗
```
border=no
script-opts-append=xxxxx
```

**Reply by 921j** [reply]
> 首先区分 mpv.conf input.conf
> 
> * input_uosc.conf 依然是一种 input.conf 不可能保存在 mpv.conf 里
> * script-opts.conf profiles.conf 也是 mpv.conf ，前两者的参数可以合并进 mpv.conf
> * `script-opts-append` 是属于 mpv 的选项；`border` 也是属于 mpv 的选项。我不理解你串联两个 mpv选项 的目的。直接分开写不行吗
> 
> ```
> border=no
> script-opts-append=xxxxx
> ```

大佬真勤快！我刚刚试出来原来直接加就行了。你提醒得对，因为我以前对 mpv 没有研究，那么在 input_uosc.conf  里可以用 include 指令创建一个自定义的 input_xxx.conf 文件的意思是吗？我希望的是两个 input 中同一个键都映射了时以自定义的为准，不知道该怎么做。

**Reply by hooke007** [reply]
> 那么在 input_uosc.conf 里可以用 include 指令创建一个自定义的 input_xxx.conf 文件的意思是吗

这个问题我就不回答了，建议先看看mpv的手册 来区分 mpv.conf input.conf。涉及的参数帖子里已经有链接了
而且 input_uosc.conf 怎么生效的我的 mpv.conf 有写

> 两个 input 

input.conf 只能用一个

**Reply by 921j** [reply]
好的，多谢解惑！只能用一个的话那就算了，只需要修改两个文件也已经比之前好很多了。

**Comment by verygoodlee** [comment]
我还是建议用单独的`~~/script-opts/xxx.conf`文件管理。
当前管理模式下，如果以 `mpv xxx.mkv --script-opts="aaa=bbb"`启动的话，`script-opts`启动参数会覆盖掉设置的脚本选项，
比如这个播放b站视频的浏览器插件就用了这个参数，
https://github.com/LuckyPuppy514/Play-With-MPV/blob/cb6cd1d70434f59a7f29d1dcb9d39d0f60974123/play-with-mpv.user.js#L1998
我已经提了修改意见了 LuckyPuppy514/Play-With-MPV#130。

开发者没注意到这些细节也挺正常，可能就会造成脚本选项全失效，以`~~/script-opts/xxx.conf`管理兼容性显然更好，不怕这种冲突

**Reply by hooke007** [reply]
我没记错的话 cmd启动的选项权级高于任何 conf

**Reply by verygoodlee** [reply]
`~~/script-opts/xxx.conf`跟`--script-opts`是隔离开的，不会覆盖掉

**Reply by hooke007** [reply]
我的意思是，这两在指定同一个脚本选项的时候，仍然 `script-opts` 优先。

这里还是列表选项的错误使用问题， 正常应该优先考虑使用 `append`

**Reply by verygoodlee** [reply]
这里也犯了同样的错误 kjtsune/embyToLocalPlayer#22，开发者普遍都没注意到这个细节，导致脚本选项全失效
